### PR TITLE
[Doctrine] Promote DQL in favor of QueryBuilder

### LINKED ISSUE #10779: Promote DQL queries instead of QueryBuilder when possible
DQL in NOWDOC strings are apparently the recommended way of building non-dynamic queries. It's recommended by some of the top Doctrine developers. See this discussion and the examples to learn more about this: https://github.com/doctrine/doctrine2/issues/7520

### PR DESCRIPTION
**Pull request by wouterj**
Fixes #10779

### CODE REVIEW

File: doctrine/associations.rst [hunk 1]
```diff
@@ -471,15 +471,17 @@ following method to the ``ProductRepository`` class::
...
-            ->getQuery()
-            ->getOneOrNullResult();
+        $entityManager = $this->getEntityManager();
+
+        $query = $entityManager->createQuery(
+            'SELECT p, c
+            FROM App\Entity\Product p
+            INNER JOIN p.category c
+            WHERE p.id = :id
+            '
```
Comment (javiereguiluz) [inline doctrine/associations.rst]: Should this quote go at the end of the previous line?

File: messenger/custom-transport.rst [hunk 2]
```diff
@@ -58,10 +58,12 @@ Here is a simplified example of a database transport::
...
             // Get a message from "my_queue"
-            $row = $this->db->createQueryBuilder()
-                ->from('my_queue')
-                ->where('delivered_at is null OR delivered_at < :redeliver_timeout')
-                ->andWhere('handled = :false')
+            $row = $this->db->createQuery(
+                    'SELECT *
+                    FROM my_queue
+                    WHERE (delivered_at is null OR delivered_at < :redeliver_timeout')
+                    AND handled = :false
```
Comment (javiereguiluz) [inline messenger/custom-transport.rst]: There's a missing closing quote at the end of this line.

File: messenger/custom-transport.rst [hunk 3]
```diff
@@ -85,13 +87,9 @@ Here is a simplified example of a database transport::
             }
 
             // Mark the message as "handled"
-            $this->db->createQueryBuilder()
-                ->update('my_queue')
-                ->setValues([
-                    'handled' => true
-                ])
-                ->where('id = :id')
+            $this->db->createQuery('UPDATE my_queue SET handled = :true WHERE id = :id')
```
Comment (javiereguiluz) [inline messenger/custom-transport.rst]: I know it's common to use things like `handled = :true` ... but then it's weird to do `->setParameter('true', true)`  Should we change this to `handled = :is_handled` and `->setParameter('is_handled', true)` ?

**Review by javiereguiluz — APPROVED** [review]
Nice changes! Thank you!

File: doctrine.rst [hunk 4]
```diff
@@ -736,36 +738,46 @@ write custom queries. Now, you can call this method on the repository::
...
 
-    public function findAllGreaterThanPrice($price): array
+    // ...
+    public function findAllGreaterThanPrice($price, $includeUnavailableProducts = false): array
     {
-        $entityManager = $this->getEntityManager();
+        // automatically knows to select Products
+        // the "p" is an alias you'll use in the rest of the query
+        $qb = $this->createQueryBuilder('p')
+            ->andWhere('p.price > :price')
```
Comment (yceruto) [inline doctrine.rst]: It's the first `where` condition, maybe we should use `->where()` instead.

File: doctrine.rst [hunk 5]
```diff
@@ -736,36 +738,46 @@ write custom queries. Now, you can call this method on the repository::
...
+            ->setParameter('price', $price)
+            ->orderBy('p.price', 'ASC')
+
+        if (!$includeUnavailableProducts) {
+            $qb->andWhere('p.available = :available')
+                ->setParameter('available', true);
```
Comment (yceruto) [inline doctrine.rst]: The parameter doesn't make much sense here.

File: messenger/custom-transport.rst [hunk 6]
```diff
@@ -58,10 +58,12 @@ Here is a simplified example of a database transport::
...
         {
             // Get a message from "my_queue"
-            $row = $this->db->createQueryBuilder()
-                ->from('my_queue')
-                ->where('delivered_at is null OR delivered_at < :redeliver_timeout')
-                ->andWhere('handled = :false')
+            $row = $this->db->createQuery(
+                    'SELECT *
+                    FROM my_queue
+                    WHERE (delivered_at is null OR delivered_at < :redeliver_timeout')
```
Comment (yceruto) [inline messenger/custom-transport.rst]: `... IS NULL ...` uppercase for consistency

File: messenger/custom-transport.rst (hunk 3 above)
Comment (yceruto) [reply messenger/custom-transport.rst]: It's odd to see a parameter for a constant value, why not use `SET handled = TRUE` directly?

**Comment by wouterj** [comment]
Thanks for the quick reviews. I've resolved all comments.

The Custom Transport section is a very weird one, as it "uses a dummy DB" with sort-of Doctrine syntax. I'll rewrite the example in a new PR (and make it just doctrine DBAL).

File: messenger/custom-transport.rst [hunk 7]
```diff
@@ -116,14 +110,15 @@ Here is a simplified example of a database transport::
...
-                    'handled' => false,
-                ]);
+            $this->db->createQuery(
+                    'INSERT INTO my_queue (id, envelope, delivered_at, handled)
+                    VALUES (:id, :envelope, :delivered_at, :handled)'
+                )
+                ->setParameter('id', $uuid)
+                ->setParameter('envelope', $encodedMessage['body'])
+                ->setParameter('delivered_at', null)
+                ->setParameter('handled', false)
```
Comment (yceruto) [inline messenger/custom-transport.rst]: same here :)

**Review by yceruto — APPROVED** [review]

**Comment by javiereguiluz** [comment]
Thank you Wouter. This is really nice!

**Pull request merged**